dimod/include/dimod/constrained_quadratic_model.h

```diff
@@ -0,0 +1,493 @@
+        }
+        varinfo_[v].lb = 0;
+        varinfo_[v].ub = 1;
+        varinfo_[v].vartype = Vartype::BINARY;
+    } else if (source == Vartype::BINARY && target == Vartype::SPIN) {
+        objective.substitute_variable(v, .5, .5);
+        for (auto& constraint : constraints_) {
+            constraint.substitute_variable(v, .5, .5);
+        }
+        varinfo_[v].lb = -1;
+        varinfo_[v].ub = +1;
+        varinfo_[v].vartype = Vartype::SPIN;
+    } else if (source == Vartype::SPIN && target == Vartype::INTEGER) {
+        // first go to BINARY, then INTEGER
+        change_vartype(Vartype::BINARY, v);
+        change_vartype(Vartype::INTEGER, v);
+    } else if (source == Vartype::BINARY && target == Vartype::INTEGER) {
+        // nothing need to change except the vartype itself
+        varinfo_[v].vartype = Vartype::INTEGER;
+    } else {
+        // todo: there are more we could support
+        throw std::logic_error("unsupported vartype change");
+    }
+}
+
+template <class bias_type, class index_type>
+Constraint<bias_type, index_type>& ConstrainedQuadraticModel<bias_type, index_type>::constraint_ref(
+        index_type c) {
+    assert(c >= 0 && static_cast<size_type>(c) < num_constraints());
+    return constraints_[c];
+}
+
+template <class bias_type, class index_type>
+const Constraint<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::constraint_ref(index_type c) const {
+    assert(c >= 0 && static_cast<size_type>(c) < num_constraints());
+    return constraints_[c];
+}
+
+template <class bias_type, class index_type>
+template <class T>
+void ConstrainedQuadraticModel<bias_type, index_type>::fix_variable(index_type v, T assignment) {
+    assert(v >= 0 && static_cast<size_type>(v) < num_variables());
+    substitute_variable(v, 0, assignment);
+    remove_variable(v);
+}
+
+template <class bias_type, class index_type>
+bias_type ConstrainedQuadraticModel<bias_type, index_type>::lower_bound(index_type v) const {
+    return varinfo_[v].lb;
+}
+
+template <class bias_type, class index_type>
+std::size_t ConstrainedQuadraticModel<bias_type, index_type>::num_constraints() const {
+    return constraints_.size();
+}
+
+template <class bias_type, class index_type>
+std::size_t ConstrainedQuadraticModel<bias_type, index_type>::num_variables() const {
+    return varinfo_.size();
+}
+
+template <class bias_type, class index_type>
+void ConstrainedQuadraticModel<bias_type, index_type>::remove_constraint(index_type c) {
+    constraints_.erase(constraints_.begin() + c, constraints_.begin() + c + 1);
+}
+
+template <class bias_type, class index_type>
+void ConstrainedQuadraticModel<bias_type, index_type>::remove_variable(index_type v) {
+    assert(v >= 0 && static_cast<size_type>(v) < num_variables());
+    for (auto& constraint : constraints_) constraint.reindex_variables(v);
+    objective.reindex_variables(v);
+    varinfo_.erase(varinfo_.begin() + v);
```

Comment:
What happens when constraints become empty due to removal of variables, do they remain in the system. This will trigger the addition of empty variables to the serverside cqm.

Reply:
Yes, it leaves empty variables in the system. This is desired. Presolve removes them later.